### ISSUE TITLE
Prototype of selective RHS ODE computation

### ISSUE THREAD

**Issue by casella** [OpenModelica/OpenModelica#11432]
The attached 
[HeatingSystemExogenous](https://github.com/OpenModelica/OpenModelica/files/13077754/HeatingSystemExogenous.mo.txt) package contains some test cases of systems that can hugely benefit from multi-rate integration, because the majority of the steps involves two-three state variables rather than 2*N.

Running the multi-rate simulations today results in a small reduction of the simulation time, because of the reduction of the time spent by the implicit solver. However, the full RHS is still computed also for local refinement steps. A more significant time saving could be achieved if only the two-three derivatives that are actually needed to compute the local refinement steps were computed, instead of the full derivative record.

I understand implementing this feature in general could be a non-trivial issue. However, it would be really nice if we could have a prototype version of it that works with such simple examples:

- purely continuous time
- index-1 system, can be turned into ODE form exclusively by means of explicit assignments
- no external functions, no need of index reduction, no dynamic state selection, no algebraic loops, no events, no discrete part, no other fancy stuff

@phannebohm how do you see that?

Keeping @bernhardbachmann in the loop.

**Comment by phannebohm** [comment]
I started working on this in #12934. It seems that evaluating the RHS-equations in a for loop instead of a hard coded list is not loosing any efficiency, so that's good. See [report](https://libraries.openmodelica.org/branches/history/master/2024-09-26%2004:14:09..2024-09-26%2017:41:33.html).

The next part is dynamically generating an appropriate subset of indices based on selection of fast states and giving that into `functionXXX`. Then we should be good to go. I''ll be working on that now.

**Comment by casella** [comment]
Excellent! This is actually one of the test cases of our multi-rate paper, so if we can get measurable advancements out of it, it would be great. 

Next will be some example involving slow variables with IF97 models. That could be even better.

**Comment by casella** [comment]
See PR #14016

**Comment by casella** [comment]
Update:
- this feature is currently available in experimental form in PR #14016, and it's been used for the revised version of the JOMP paper
- it should be completed (see e.g. #14072) and merged on master
- as I understand, for a system with N states and M << N fast states, the current implementation is quite fast, but still has O(N) complexity. Maybe this is not a problem, but in case profiling reveals that this still takes a significant amount of time, we could consider implementations do not scan the entire RHS, but only do a search of what is needed to compute each selected fast state. This could end up having O(M^2) complexity, which could be much better than O(N) for M << N. Eg. N =1000, M = 5, 25 << 1000

**Comment by phannebohm** [comment]
>   * as I understand, for a system with N states and M << N fast states, the current implementation is quite fast, but still has O(N) complexity. Maybe this is not a problem, but in case profiling reveals that this still takes a significant amount of time, we could consider implementations do not scan the entire RHS, but only do a search of what is needed to compute each selected fast state. This could end up having O(M^2) complexity, which could be much better than O(N) for M << N. Eg. N =1000, M = 5, 25 << 1000

It will probably be O(E log E) where E is the number of equations selected, which may be significantly larger than M if the system has many dependencies or a "bottleneck" where some equations need to be evaluated always. But if that's too slow then multi-rate may not be the best choice for those models.

Also the current implementation introduces a lot of generated code for the dependency graph which is static. I guess this is not what we want as it would introduce performance regressions similar to #13999. I will try to move that part to the backend and write it to a binary file like we did in #10352 for Jacobian sparsity patterns.

**Comment by casella** [comment]
> It will probably be O(E log E) where E is the number of equations selected, which may be significantly larger than M if the system has many dependencies or a "bottleneck" where some equations need to be evaluated always. But if that's too slow then multi-rate may not be the best choice for those models.

Exactly. I think if the system is good for multirate, E will have the same order of magnitude than M
 
> Also the current implementation introduces a lot of generated code for the dependency graph which is static. I guess this is not what we want as it would introduce performance regressions similar to [#13999](https://github.com/OpenModelica/OpenModelica/issues/13999). I will try to move that part to the backend and write it to a binary file like we did in [#10352](https://github.com/OpenModelica/OpenModelica/pull/10352) for Jacobian sparsity patterns.

Sounds good.

**Comment by casella** [comment]
@phannebohm is now working to get the structural information from the json file instead of embedding it in the C code.

**Comment by casella** [comment]
PR #14016 is now merged onto master. @phannebohm, is there anything still missing?

**Comment by casella** [comment]
@phannebohm please also consider
- #14904
- #14983
- #15205

Thanks!
